lib/base/namespace.cpp

```diff
@@ -103,111 +122,45 @@ void Namespace::Freeze() {
 	m_Frozen = true;
 }
 
-void Namespace::RemoveAttribute(const String& field)
-{
-	ObjectLock olock(this);
-
-	Namespace::Iterator it;
-	it = m_Data.find(field);
-
-	if (it == m_Data.end())
-		return;
-
-	m_Data.erase(it);
-}
-
-NamespaceValue::Ptr Namespace::GetAttribute(const String& key) const
-{
-	ObjectLock olock(this);
-
-	auto it = m_Data.find(key);
-
-	if (it == m_Data.end())
-		return nullptr;
-
-	return it->second;
-}
-
-void Namespace::SetAttribute(const String& key, const NamespaceValue::Ptr& nsVal)
+std::shared_lock<std::shared_timed_mutex> Namespace::ReadLockUnlessFrozen() const
 {
-	ObjectLock olock(this);
-
-	m_Data[key] = nsVal;
+	if (m_Frozen.load(std::memory_order_relaxed)) {
+		return std::shared_lock<std::shared_timed_mutex>();
+	} else {
+		return std::shared_lock<std::shared_timed_mutex>(m_DataMutex);
+	}
 }
 
 Value Namespace::GetFieldByName(const String& field, bool, const DebugInfo& debugInfo) const
 {
-	ObjectLock olock(this);
+	auto lock (ReadLockUnlessFrozen());
 
-	auto nsVal = GetAttribute(field);
+	auto nsVal = m_Data.find(field);
 
-	if (nsVal)
-		return nsVal->Get(debugInfo);
+	if (nsVal != m_Data.end())
+		return nsVal->second.m_Value;
 	else
 		return GetPrototypeField(const_cast<Namespace *>(this), field, false, debugInfo); /* Ignore indexer not found errors similar to the Dictionary class. */
 }
 
 void Namespace::SetFieldByName(const String& field, const Value& value, bool overrideFrozen, const DebugInfo& debugInfo)
 {
-	ObjectLock olock(this);
-
-	auto nsVal = GetAttribute(field);
+	// The override frozen parameter is mandated by the interface but ignored here. If the namespace is frozen, this
+	// disables locking for read operations, so it must not be modified again to ensure the consistency of the internal
+	// data structures.
+	(void) overrideFrozen;
```

Comment:
I prefer my way, see #9411.

Reply:
But then then the parameter would have no name and the comment would have to describe it explicitly. "The unnamed parameter of this function is the override frozen parameter mandated by the interface". Then why not just give it a name?

Reply:
No name – no comment needed for the name as there's no name. 😎

Reply:
But the comment also has a purpose. If at some time in the future, someone thinks it might be a good idea to support `overrideFrozen` here, they will rather soon find this comment, rather than trying to start implementing it and figuring this out themselves ("just" wasted time) or worse, don't figure it out and introduce a race condition. So please don't ask me to remove (correct) documentation.